contracts/adapters/swap/drop/src/contract.rs

```diff
@@ -0,0 +1,320 @@
+            };
+
+            to_json_binary(&SimulateSwapExactAssetInResponse {
+                asset_out: Asset::Native(Coin::new(
+                    (exchange_rate * asset_in.amount()).into(),
+                    asset_out_denom,
+                )),
+                spot_price,
+            })
+        }
+        QueryMsg::SimulateSwapExactAssetOutWithMetadata {
+            asset_out,
+            include_spot_price,
+            ..
+        } => {
+            let asset_in_denom =
+                get_opposite_denom(asset_out.denom(), &bridged_denom, &canonical_denom);
+
+            let exchange_rate = get_exchange_rate(deps)?;
+
+            let spot_price = if include_spot_price {
+                Some(exchange_rate)
+            } else {
+                None
+            };
+
+            to_json_binary(&SimulateSwapExactAssetOutResponse {
+                asset_in: Asset::Native(Coin::new(
+                    (asset_out.amount().div_floor(exchange_rate)).into(),
+                    asset_in_denom,
+                )),
+                spot_price,
+            })
+        }
+        QueryMsg::SimulateSmartSwapExactAssetIn { asset_in, .. } => {
+            let asset_out_denom =
+                get_opposite_denom(asset_in.denom(), &bridged_denom, &canonical_denom);
+
+            let exchange_rate = get_exchange_rate(deps)?;
+
+            to_json_binary(&Asset::Native(Coin::new(
+                (exchange_rate * asset_in.amount()).into(),
+                asset_out_denom,
+            )))
+        }
+        QueryMsg::SimulateSmartSwapExactAssetInWithMetadata {
+            asset_in,
+            include_spot_price,
+            ..
+        } => {
+            let asset_out_denom =
+                get_opposite_denom(asset_in.denom(), &bridged_denom, &canonical_denom);
+
+            let exchange_rate = get_exchange_rate(deps)?;
+
+            let spot_price = if include_spot_price {
+                Some(exchange_rate)
+            } else {
+                None
+            };
+
+            to_json_binary(&SimulateSwapExactAssetInResponse {
+                asset_out: Asset::Native(Coin::new(
+                    (exchange_rate * asset_in.amount()).into(),
+                    asset_out_denom,
+                )),
+                spot_price,
+            })
+        }
+    }
+    .map_err(From::from)
+}
+
+fn get_opposite_denom(denom: &str, bridged_denom: &str, canonical_denom: &str) -> String {
+    match denom {
+        denom if denom == bridged_denom => canonical_denom.to_string(),
+        denom if denom == canonical_denom => bridged_denom.to_string(),
+        _ => unimplemented!(),
+    }
+}
+
+fn get_exchange_rate(deps: Deps) -> ContractResult<Decimal> {
+    let drop_core_contract_address = DROP_CORE_CONTRACT_ADDRESS.load(deps.storage)?;
+
+    let exchange_rate: Decimal = deps.querier.query_wasm_smart(
+        drop_core_contract_address,
+        &drop_staking_base::msg::core::QueryMsg::ExchangeRate {},
```

Comment:
@albertandrejev How accurate is this for arbitrary bond sizes? Let's say exchange rate returns 1:1, will the realized exchanged amount if put in 1 unit of X vs 1_000_000_000 units of X be the same? If not, is there a way to simulate the actual outcome which takes into account bond size?

Reply:
Yes, exchange rate is linear in frames of one transaction, exchanged amount will be with same exchange rate inside  one message.

Reply:
No need to simulate

Reply:
Awesome, thank you!